File: src/VideoEffectsSdkReactNativeModule.ts
import { requireNativeModule, type EventSubscription } from "expo-modules-core";
import type { NativeModule } from "expo-modules-core/build/ts-declarations/NativeModule";

import type {
  BlurOptions,
  EffectsConfig,
  EffectsEvent,
  EffectsState,
  EffectsUnavailableReason,
  FrameCaptureEvent,
  InitializationResult,
  NativeModuleEventsMap,
  NativeModuleInterface,
  ReplaceOptions,
  SegmentationPreset,
} from "./VideoEffectsSdkReactNativeModule.types";

const VideoEffectsNativeModule = requireNativeModule<
  NativeModule<NativeModuleEventsMap> & NativeModuleInterface
>("VideoEffectsSdkReactNativeModule");

class TsvbVideoEffects {
  private _state: EffectsState = {
    isInitialized: false,
    isReady: false,
    activeEffect: "none",
    isEffectsUnavailable: false,
    error: null,
  };
  private _subscribers = new Set<(event: EffectsEvent) => void>();
  private _frameCaptureSubscription: EventSubscription | null = null;

  constructor() {
    // Eager at module load so the listener is live before any stall and survives cleanup()→re-init.
    // Not stored — it lives for the process lifetime and the native emitter holds the callback.
    VideoEffectsNativeModule.addListener("onEffectsUnavailable", ({ reason }) =>
      this.handleEffectsUnavailable(reason),
    );
  }

  async initialize(config: EffectsConfig): Promise<InitializationResult> {
    if (this._state.isEffectsUnavailable) {
      throw new Error(this._state.error ?? "Effects SDK unavailable");
    }
    if (this._state.isInitialized) {
      return { success: true, status: "already_initialized" };
    }

    const { trackId } = config;

    try {
      const result = await VideoEffectsNativeModule.initialize(config.customerID, trackId);

      if (!result.success) {
        const msg = result.error || "Initialization failed";
        this.updateState({ error: msg, isEffectsUnavailable: true });
        throw new Error(msg);
      }

      this.updateState({
        isInitialized: true,
        isReady: true,
        error: null,
      });

      return result;
    } catch (error) {
      const msg = `Failed to initialize TSVB SDK: ${error}`;
      if (!this._state.isEffectsUnavailable) {
        this.updateState({ error: msg, isEffectsUnavailable: true });
      }
      throw new Error(msg);
    }
  }

  async enableBlur(options?: BlurOptions): Promise<void> {
    this.ensureInitialized();
    this.ensureEffectsAvailable();

    try {
      const power = options?.power ?? 0.5;
      await VideoEffectsNativeModule.enableBlurBackground(power);
      this.updateState({ activeEffect: "blur", error: null });
    } catch (error) {
      const msg = `Failed to enable blur: ${error}`;
      this.emitError(msg, true);
      throw new Error(msg);
    }
  }

  async enableReplaceBackground(options: ReplaceOptions): Promise<void> {
    this.ensureInitialized();
    this.ensureEffectsAvailable();

    try {
      await VideoEffectsNativeModule.enableReplaceBackground(options.image);
      this.updateState({ activeEffect: "replace", error: null });
    } catch (error) {
      const msg = `Failed to enable background replacement: ${error}`;
      this.emitError(msg, true);
      throw new Error(msg);
    }
  }

  async disableEffects(): Promise<void> {
    this.ensureInitialized();

    try {
      if (this._state.activeEffect === "blur") {
        await VideoEffectsNativeModule.disableBlurBackground();
      } else if (this._state.activeEffect === "replace") {
        await VideoEffectsNativeModule.disableReplaceBackground();
      }
      this.updateState({ activeEffect: "none", error: null });
    } catch (error) {
      const msg = `Failed to disable effects: ${error}`;
      this.emitError(msg, true);
      throw new Error(msg);
    }
  }

  getState(): EffectsState {
    return { ...this._state };
  }

  subscribe(callback: (event: EffectsEvent) => void): () => void {
    this._subscribers.add(callback);
    return () => {
      this._subscribers.delete(callback);
    };
  }

  /** Set segmentation quality preset. Only effective on iOS — Android handles this internally. */
  setSegmentationPreset(preset: SegmentationPreset): void {
    VideoEffectsNativeModule.setSegmentationPreset(preset);
  }

  /**
   * Start periodic frame capture. Captured frames are saved as JPEG files
   * and emitted via the subscriber callback as `frameCaptured` events.
   * @param intervalMs Capture interval in milliseconds (default: 5000)
   */
  startFrameCapture(intervalMs: number = 5000): void {
    this.ensureInitialized();

    if (!this._frameCaptureSubscription) {
      this._frameCaptureSubscription = VideoEffectsNativeModule.addListener(
        "onFrameCaptured",
        (event: FrameCaptureEvent) => {
          this.emit({ type: "frameCaptured", frame: event });
        },
      );
    }

    VideoEffectsNativeModule.startFrameCapture(intervalMs);
  }

  /** Stop periodic frame capture. */
  stopFrameCapture(): void {
    VideoEffectsNativeModule.stopFrameCapture();
    this._frameCaptureSubscription?.remove();
    this._frameCaptureSubscription = null;
  }

  cleanup(): void {
    this.stopFrameCapture();

    try {
      VideoEffectsNativeModule.cleanup();
    } catch (error) {
      // Surface to subscribers but don't throw — cleanup must always complete.
      const msg = error instanceof Error ? error.message : String(error);
      this.emitError(`TSVB native cleanup failed: ${msg}`, true);
    }

    this._state = {
      isInitialized: false,
      isReady: false,
      activeEffect: "none",
      isEffectsUnavailable: false,
      error: null,
    };
    this.emit({ type: "stateChange", state: this.getState() });
  }

  // --- Private ---

  /** Query native for fallback state and update local state. Returns true if effects are unavailable. */
  checkEffectsAvailability(): boolean {
    try {
      const unavailable = VideoEffectsNativeModule.isEffectsUnavailable();
      if (unavailable !== this._state.isEffectsUnavailable) {
        this.updateState({ isEffectsUnavailable: unavailable });
      }
      return unavailable;
    } catch {
      return false;
    }
  }

  private ensureInitialized(): void {
    if (!this._state.isInitialized) {
      throw new Error("TSVB SDK is not initialized. Call initialize() first.");
    }
  }

  private ensureEffectsAvailable(): void {
    if (this.checkEffectsAvailability()) {
      throw new Error(
        "Effects unavailable — camera is running in fallback mode without effects pipeline."
      );
    }
  }

  /** Bridges the native fallback push into local state + a typed app-facing event. Idempotent. */
  private handleEffectsUnavailable(reason: EffectsUnavailableReason): void {
    if (this._state.isEffectsUnavailable) {
      return;
    }
    // updateState auto-emits stateChange for getState() readers; emit delivers the typed reason.
    this.updateState({
      isEffectsUnavailable: true,
      error: `Effects unavailable (${reason})`,
    });
    this.emit({ type: "effectsUnavailable", reason });
  }

  private updateState(partial: Partial<EffectsState>): void {
    this._state = { ...this._state, ...partial };
    this.emit({ type: "stateChange", state: this.getState() });
  }

  private emitError(error: string, recoverable: boolean): void {
    this.emit({ type: "error", error, recoverable });
  }

  private emit(event: EffectsEvent): void {
    this._subscribers.forEach(cb => {
      try {
        cb(event);
      } catch {
        // Don't let subscriber errors propagate
      }
    });
  }
}

export const tsvbVideoEffects = new TsvbVideoEffects();

export * from "./VideoEffectsSdkReactNativeModule.types";
export { TsvbVideoEffects };
export { VideoEffectsNativeModule as TsvbNativeModule };

File: android/src/main/java/expo/modules/videoeffectssdkreactnative/VideoEffectsSdkReactNativeModule.kt
package expo.modules.videoeffectssdkreactnative

import android.content.Context
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

class VideoEffectsSdkReactNativeModule : Module() {

    private val tsvbManager by lazy {
        TsvbManager(context)
    }

    private val context: Context
        get() = appContext.reactContext ?: throw IllegalStateException("React context not available")

    override fun definition() = ModuleDefinition {
        Name("VideoEffectsSdkReactNativeModule")

        Events("onFrameCaptured")

        AsyncFunction("initialize") { customerID: String, trackId: String, promise: Promise ->
            tsvbManager.initialize(customerID, trackId) { result ->
                promise.resolve(result)
            }
        }

        AsyncFunction("enableBlurBackground") { power: Double?, promise: Promise ->
            val blurPower = power ?: 0.5
            tsvbManager.enableBlurBackground(blurPower.toFloat()) { result ->
                if (result["success"] == true) {
                    promise.resolve(result)
                } else {
                    promise.reject("EFFECTS_ERROR", result["error"] as? String ?: "Unknown error", null)
                }
            }
        }

        AsyncFunction("disableBlurBackground") { promise: Promise ->
            tsvbManager.disableBlurBackground { result ->
                promise.resolve(result)
            }
        }

        AsyncFunction("enableReplaceBackground") { assetSource: Map<String, Any>?, promise: Promise ->
            tsvbManager.enableReplaceBackground(assetSource) { result ->
                if (result["success"] == true) {
                    promise.resolve(result)
                } else {
                    promise.reject("EFFECTS_ERROR", result["error"] as? String ?: "Unknown error", null)
                }
            }
        }

        AsyncFunction("disableReplaceBackground") { promise: Promise ->
            tsvbManager.disableReplaceBackground { result ->
                promise.resolve(result)
            }
        }

        Function("isBlurEnabled") {
            tsvbManager.isBlurEnabled
        }

        Function("hasVirtualBackground") {
            tsvbManager.isReplaceBackgroundEnabled
        }

        Function("isInitialized") {
            tsvbManager.isInitialized
        }

        Function("isEffectsUnavailable") {
            tsvbManager.isEffectsUnavailable
        }

        Function("cleanup") {
            tsvbManager.cleanup()
        }

        Function("setDeviceOrientation") { _: String ->
            // No-op on Android — orientation is handled by CameraPipeline internally
        }

        Function("setSegmentationPreset") { _: String ->
            // No-op on Android — SDK uses SegmentationMode (orientation-based), not quality presets
        }

        Function("startFrameCapture") { intervalMs: Int ->
            tsvbManager.startFrameCapture(intervalMs.toLong()) { filePath, width, height, timestamp ->
                sendEvent("onFrameCaptured", mapOf(
                    "filePath" to filePath,
                    "timestamp" to timestamp,
                    "width" to width,
                    "height" to height,
                ))
            }
        }

        Function("stopFrameCapture") {
            tsvbManager.stopFrameCapture()
        }
    }
}
